presto-main/src/main/java/com/facebook/presto/operator/LookupJoinPageBuilder.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.List;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This page builder creates pages with dictionary blocks:
+ * normal dictionary blocks for the probe side and the original blocks for the build side.
+ *
+ * TODO use dictionary blocks (probably extended kind) to avoid data copying for build side
+ */
+public class LookupJoinPageBuilder
+{
+    private final IntArrayList probeIndexBuilder = new IntArrayList();
+    private final PageBuilder buildPageBuilder;
+    private final int buildOutputChannelCount;
+    private int probeBlockBytes;
+
+    public LookupJoinPageBuilder(List<Type> buildTypes)
+    {
+        this.buildPageBuilder = new PageBuilder(requireNonNull(buildTypes, "buildTypes is null"));
+        this.buildOutputChannelCount = buildTypes.size();
+    }
+
+    public boolean isFull()
+    {
+        return probeBlockBytes + buildPageBuilder.getSizeInBytes() >= DEFAULT_MAX_PAGE_SIZE_IN_BYTES || buildPageBuilder.isFull();
+    }
+
+    public boolean isEmpty()
+    {
+        return probeIndexBuilder.isEmpty() && buildPageBuilder.isEmpty();
+    }
+
+    public void reset()
+    {
+        // be aware that probeIndexBuilder will not clear its capacity
+        probeIndexBuilder.clear();
+        buildPageBuilder.reset();
+        probeBlockBytes = 0;
+    }
+
+    /**
+     * append the index for the probe and copy the row for the build
+     */
+    public void appendRow(JoinProbe probe, LookupSource lookupSource, long joinPosition)
+    {
+        // probe side
+        appendProbeIndex(probe);
+
+        // build side
+        buildPageBuilder.declarePosition();
+        lookupSource.appendTo(joinPosition, buildPageBuilder, 0);
+    }
+
+    /**
+     * append the index for the probe and append nulls for the build
+     */
+    public void appendNullForBuild(JoinProbe probe)
+    {
+        // probe side
+        appendProbeIndex(probe);
+
+        // build side
+        buildPageBuilder.declarePosition();
+        for (int i = 0; i < buildOutputChannelCount; i++) {
+            buildPageBuilder.getBlockBuilder(i).appendNull();
+        }
+    }
+
+    public Page build(JoinProbe probe)
+    {
+        Block[] blocks = new Block[probe.getOutputChannelCount() + buildOutputChannelCount];
+        int[] probeIndices = probeIndexBuilder.toIntArray();
+        verify(buildPageBuilder.getPositionCount() == probeIndices.length);
+
+        int[] probeOutputChannels = probe.getOutputChannels();
```

Comment:
```
if (probeIndices are a simple covering of the block)
   copy block reference
else if (probeIndices are sequential without holes) 
  call get region on the blocks instead of mask
else 
  mask the blocks
``